### ISSUE TITLE
fit_predict

### ISSUE THREAD

**Issue by jasmainak** [pavanramkumar/pyglmnet#37]
easy fix: if you remove `fit_param` from here: https://github.com/pavanramkumar/pyglmnet/blob/master/pyglmnet/pyglmnet.py#L294 (you can already use `self.fit_param`), then you can have a `fit_predict` convenience method which just returns `self.fit(X, y).predict(X)`

**Comment by titipata** [comment]
The thing is @pavanramkumar put list of regularized regression parameters in `reg_lambda` so that turns out to give list of fitted parameters back i.e. `{['beta0': ..., 'beta': ...], [...], [...]}`. Should we only receive one parameter at a time (`self.reg_lambda`) and save output to `self.beta_`?

**Comment by jasmainak** [comment]
Ah, is this due to the warm start to have faster computation? Since it's the outermost loop, you can abstract it away from the `fit` method. I think the way `scikit-learn` does it -- it uses the `self.beta_` from the last call to the method for the warm start. So, I would suggest something like:
```py
if hasattr(self, 'beta_'):
    beta = self.beta_
else:
   beta = np.random(...)
```
and then continue as usual without the loop.

**Comment by pavanramkumar** [comment]
@jasmainak yes, it's for warm restarts! I like the suggestion of abstracting away the fitting algorithm from looping through the regularization path.

But I would like it to mimic the functionality of the R package which returns a list of fits for the specified regularization params in `reg_lambda`.

I would also like to adopt a practice which easily allows future modification to automatically fit the best `reg_lambda` by cross-validation.

What's the best solution? Two functions? `path_fit()` and `fit()`

**Comment by jasmainak** [comment]
API design is hard :) The best way to write good API is to write down an example how you will use it. If you can share 5 lines of boiler-plate code to demonstrate how you plan to use it, that would be really cool. Do you want to be able to fit on multiple lambdas and predict on one (for efficient grid search?) Or fit on multiple and predict on a subset (YAGNI?). Or fit on one lambda and predict one one?

**Comment by titipata** [comment]
I vote for predict multiple lambdas and predict/store one best parameters in `self.beta_` and `fit_predict` will use `self.beta` to predict. Rather that that, we stores all parameters in `self.fit_params` i.e. `fit_params = [{'beta': ..., 'reg_lambda': ...}, ...]`.

**Comment by jasmainak** [comment]
ok, I did some thinking. We could hack the magic method `__getitems__` of the object. So, you have `self.betas_` as you have now which is a list / dict for different lambdas. But when the user tries to slice it, you get an object which has only the subset of betas that the user asked for. E.g.,

```py
glms = GLMPath(reg_lambdas)  # it's a list-like object
glms.fit(X_train, y_train)  # you fit only lambdas
glms[3].fit_predict(X, y)  # predict on only one
devs = glms[2:5].score(X_test, y_test, method='deviance')
r2s = glms[2:5].score(X_test, y_test, method='r2')
glms.plot_path()  # plot the regularization path
```

**Comment by hugoguh** [comment]
@titipata with your suggestion I’m a bit concerned about overfitting. The chosen lambda should have best performance on test set. So the best at doing `self.fit(Xr, y).predict(Xt)`

**Comment by pavanramkumar** [comment]
@jasmainak great advice! 

Just looked up the `__getitem__` option. Sounds really clever but I'm not completely clear on whether the slicing is compatible with warm restarts. 

Would `GLMPath()` create a list of `GLM()` instances? If so, once `glms[0]` is fit, is it easy to initialize `glms[1]`?

Alternatively, I'd like to keep the flavor as close to the R package as possible. Here's a suggestion without the fancy slicing (comes very close to what @titpata proposed):

- Fit on multiple lambda's with warm restarts:
```
model = GLM()
model.fit(Xr, yr, reg_lambda)
```
will update `self.fit_` which is a list of dictionaries, one for each entry of `reg_lambda`, just as it is now

- Find the best lambda:

`self._cv_reg_lambda` , a callable from within `self.fit` will find the best lambda based on minimizing `self.deviance` and update  `self.beta0_` and `self.beta_`

- Simplify the `predict()` method:

it doesn't need a `fit_param` input anymore, it just predicts based on `self.beta0_` and `self.beta`

- `fit_predict(Xr, yr, Xt)` is simply `self.fit(Xr, yr).predict(Xt)`

- @hugoguh yes, having a validation set is overfitting proof but it's important not to force the user to provide one. we can implement `self.fit` and `self._cv_reg_lambda` such that user can optionally provide a validation set (default would be to use the training set alone, as is done in the R package).

**Comment by jasmainak** [comment]
@pavanramkumar:

On Thursday, 5 May 2016, pavanramkumar <notifications@github.com> wrote:

> @jasmainak <https://github.com/jasmainak> great advice!
>
> Just looked up the __getitem__ option. Sounds really clever but I'm not
> completely clear on whether the slicing is compatible with warm restart
>
Of course it would :) The fit method will do warm restarts if reg_lambda is
a list

> Would GLMPath() create a list of GLM() instances?
>
I just called it differently since it was working differently. So, no it
won't create a list of GLM() instances

> If so, once glms[0] is fit, is it easy to initialize glms[1]?
>
The current class we have will have an extra __getitem__ method that's it
:)

> Alternatively, I'd like to keep the flavor as close to the R package as
> possible. Here's a suggestion without the fancy slicing (comes very close
> to what @titpata proposed):
>
>    - Fit on multiple lambda's with warm restarts:
>
> model = GLM()
> model.fit(Xr, yr, reg_lambda)
>
> will update self.fit_ which is a list of dictionaries, one for each entry
> of reg_lambda, just as it is now
>
>    - Find the best lambda:
>
> self._cv_reg_lambda , a callable from within self.fit will find the best
> lambda
>
Is this something that the user has an option of changing. I would try to
keep cross validation options flexible

**Comment by hugoguh** [comment]
@pavanramkumar I wasn’t saying the user should give the validation test, just pointing out the dangers of overfitting if fit method returns only the best fit on training set. 
Won’t it be easier/natural to code the `cv.pyglmnet` it the `fit_predict` outputs the whole array

@jasmainak I agree with keeping the options flexible.
I like how R returns all the cv values for the whole set/path of lambdas. It is useful for 2 things (at least): 
1. to have an idea of how sensitive the result is the value of lambda
2. often the range of default or given lambdas doesn’t cover the minimum value, and looking at the whole array of `cv.fit` values is how you spot that

**Comment by pavanramkumar** [comment]
ok so let's keep it simple:

- with `__getitem__()` we can have the flexibility to fit and predict for any reg_lambda
- `fit()` will update `self.fit_` as it currently does and store a list of dicts
- `predict()` and `fit_predict()` will call predict on each lambda by default, or a subset of lambda's as sliced by the user.
- @jasmainak do you want to code this?
 
For cross-validation:
- A new class `cvpyglmnet` will inherit some methods from `pyglmnet` and provide methods for cross-validation, visualization of regularization path, etc. 
- This will preserve the R package flavor as much as possible. Ideally this just lives in a new script 
- `cvpyglmnet.py` 
- @hugoguh  do you want to code this?

**Comment by jasmainak** [comment]
Yes, I can make a PR soon enough :)

About cross validation, have you thought about using sklearn cross
validation objects? Basically, the only thing you need to recode is how to
parallelize it using joblibs. Instead of parallelizing over lambdas you
will do it across folds

**Comment by hugoguh** [comment]
@pavanramkumar yes please!
we should at some point discuss if there is a plan for having optimization for alpha also. 

side issue, regarding the warm starts, should check if the only thing that changes with warm starts is the execution time.
